### DISCUSSION TITLE
Using events to override insert()

### DISCUSSION THREAD

**Discussion by Tobeyforce** [sqlalchemy/sqlalchemy#11389] [Usage Questions]
I'm working on a multi tenant application and have desperately been trying to enforce the usage of tenant id's on all queries made in the application. It seems based on what I've read so far that the events, e.g do_orm_execute is the way to go since 2.0. Right now I'm attempting to cover select(), delete(), update(), insert() from the core using the do_orm_execute event. However, it seems like I'll also need to find ways to intercept the unit-of-work-style methods from the session .add(), delete(), update() (which might be a separate discussion, even though I'm highly curious on how to do that consistently)

For standard **selects** i've been using the with_loader_critera:

```
    if execute_state.is_select:
        # Retrieve tenant_id outside of the lambda to avoid caching issues
        tenant_id = execute_state.session.info.get("tenant_id")
        if tenant_id is not None:
            # Use a stable reference within the lambda to ensure consistent caching
            execute_state.statement = execute_state.statement.options(
                with_loader_criteria(
                    Base,  # Ensure this targets all relevant models
                    lambda cls, tenant_id=tenant_id: cls.tenant_id == tenant_id
                    if hasattr(cls, "tenant_id")
                    else None,
                    propagate_to_loaders=True,
                    include_aliases=True,
                    track_closure_variables=False,  # Optionally disable tracking if issues persist
                )
            )
```

For **delete's** or **updates** I'm trying this (update looks identical more or less):

```
@event.listens_for(Session, "do_orm_execute")
def apply_tenant_filter_on_delete(execute_state: ORMExecuteState):
    if execute_state.session.info.get("disable_events", False):
        logger.info("Skipping delete-event due to disable_events flag.")
        return  # Skip the event logic if disable_events is set

    safe = execute_state.execution_options.get("safe", True)
    if safe is False:
        logger.info("Skipping tenant filter due to safe mode.")
        return

    if execute_state.is_delete:
        tenant_id = execute_state.session.info.get("tenant_id")
        if tenant_id is not None:
            execute_state.statement = execute_state.statement.where(
                execute_state.statement.table.c.tenant_id == tenant_id
            )
```

However, I'm lost as to how to approach **inserts**. Is there any decent approach here?

The is_insert attribute indeed exists, I'm just quite lost as to how to automatically add the tenant_id-key & value to the values(). Currently I always add the tenant_id to the session, so I can grab it from session.info, the challenge is doing something like this:

```
if execute_state.is_insert:
    tenant_id = execute_state.session.info.get("tenant_id")
    # Set the tenant_id in the insert
```

**Edit 1:**
This seems to work, at least on single inserts:

```
    if execute_state.is_insert:
        tenant_id = execute_state.session.info.get("tenant_id")
        logger.debug(f"Applying tenant_id on insert: {tenant_id}")
        if tenant_id is not None:
            statement = execute_state.statement
            if hasattr(statement, "values"):
                # Create a new dictionary with the modified values
                new_values = {**statement._values, "tenant_id": tenant_id}
                execute_state.statement = statement.values(new_values)
                logger.debug(f"Modified insert statement values: {new_values}")

```
Edit 2 - This seems to support bulk inserts or single inserts with a dictionary:
E.g session.execute(insert(Course), {...})
or session.execute(insert(Course), [{}, {}])
and as previously mentioned, using .values()
session.execute(insert(Course).values(...))
```

@event.listens_for(Session, "do_orm_execute")
def apply_tenant_id_on_insert(execute_state: ORMExecuteState):
    if execute_state.session.info.get("disable_events", False):
        logger.info("Skipping event due to disable_events flag.")
        return

    safe = execute_state.execution_options.get("safe", True)
    if safe is False:
        logger.info("Skipping tenant filter due to safe mode.")
        return

    if execute_state.is_insert:
        tenant_id = execute_state.session.info.get("tenant_id")
        logger.debug(f"Applying tenant_id on insert: {tenant_id}")
        if tenant_id is not None:
            statement = execute_state.statement

            # Check if it's a bulk insert
            if execute_state.parameters:
                if isinstance(execute_state.parameters, list):
                    for param in execute_state.parameters:
                        param["tenant_id"] = tenant_id
                        logger.debug(f"Modified bulk insert parameters: {param}")
                elif isinstance(execute_state.parameters, dict):
                    # Single insert with parameters provided separately
                    execute_state.parameters["tenant_id"] = tenant_id
                    logger.debug(
                        f"Modified insert parameters: {execute_state.parameters}"
                    )
            else:
                # Single insert with values in the statement
                if hasattr(statement, "values"):
                    new_values = {**statement._values, "tenant_id": tenant_id}
                    execute_state.statement = statement.values(new_values)
                    logger.debug(f"Modified insert statement values: {new_values}")
```

Feedback?

**Comment by zzzeek** [comment]
hi -

having fields "required" for an INSERT means you just make sure the target column is NOT NULL in the database.    an INSERT that fails to include a value for that column will fail.    your request is "enforce the usage of tenant ids" so "enforce" usually means, "disallow it to be absent, i.e. fail".    if someone is coding an insert() statement to pass directly to session.execute(), that's a special case for the ORM which one does if they are doing [bulk inserts](https://docs.sqlalchemy.org/en/20/orm/queryguide/dml.html#orm-bulk-insert-statements) so it's reasonable that they need to include the tenant id as well in their statement.

Also, for the ORM-enabled delete() and update(), the tenant id can be included in the same way as you do for select(), using the `with_loader_criteria()` option.   See https://docs.sqlalchemy.org/en/20/orm/queryguide/dml.html#important-notes-and-caveats-for-orm-enabled-update-and-delete .    so you can have that approach for select(), update(), delete() that are sent to session.execute() equally.   (noting it only takes place for ORM-enabled statements, not those that are against plain `Table` objects).

For unit of work, which is `session.add()`, again it's only INSERTs where you might have an object that lacks a tenant id so the NOT NULL thing would prevent that from being missing.    for any other unit of work operation, you are dealing with a specific already-existing primary key that you would have gotten from a SELECT.

All of the above covers, "make sure a tenant_id is present, if not fail".  now if you are trying to enforce that all tenant_ids within a transaction **match** a per-request transaction id, and fail if they dont match, that's a different level of enforcement.   For that kind of thing I might want to look at PG's row-level security policies (https://www.postgresql.org/docs/current/ddl-rowsecurity.html) - this typically assigns per-row permissions to individual database users, however you'd want to use a session/transaction-level setting for this rather than individual database logins - I found an interesting blog post on how to do this at https://aws.amazon.com/blogs/database/multi-tenant-data-isolation-with-postgresql-row-level-security/ .   if I were really worried about tenant security I would see if I could make that happen at the DB level, though I dont have direct experience using this feature.

**Reply by Tobeyforce** [reply]
That really helps. 
I suppose my use case is that I'm trying to create an abstraction layer as to prevent developers from messing up, in such a way that the tenant_id should be set automatically using the current users tenant_id, which is fetched from the database (and therefore should never have to be provided by the client). However, providing a tenant_id manually should also be possible, I think. Not sure what would be the best approach here, the fundamental idea is prevention of using the wrong tenant_id, or not using any tenant_id at all when performing queries.

I'm currently removing the delete and update events for the core methods based on your feedback, and instead trying to use with_loader_critera. However, it seems like with_loader_critera doesn't capture e.g bulk updates. I'm curious if I'm missing something, or if it's not supported?:

```
    if execute_state.is_select or execute_state.is_update or execute_state.is_delete:
        # Retrieve tenant_id outside of the lambda to avoid caching issues
        tenant_id = execute_state.session.info.get("tenant_id")
        if tenant_id is not None:
            # Use a stable reference within the lambda to ensure consistent caching
            execute_state.statement = execute_state.statement.options(
                with_loader_criteria(
                    Base,  # Ensure this targets all relevant models
                    lambda cls, tenant_id=tenant_id: cls.tenant_id == tenant_id
                    if hasattr(cls, "tenant_id")
                    else None,
                    propagate_to_loaders=True,
                    include_aliases=True,
                    track_closure_variables=False,  # Optionally disable tracking if issues persist
                )
            )
```

As for enforcing the tenant_id on a db-layer during inserts, I think it makes sense that I do not allow it to be null, but I still want to make sure it's applied automatically (I guess the second best thing is to have the not null constraint raised)

And as you mentioned, I did dabble with the unit-of-work parts, I always have to use a select prior, which the with_loader_critera covers nicely, so that's great. 
For this reason I'm guessing the delete-part in this before_flush event might be completely unnecessary. 
```
event.listens_for(Session, "before_flush")
def set_tenant_id_on_add_and_validate_delete(
    session: Session, flush_context, instances
):
    """
    This event listener sets the tenant_id on new instances if it's not already set,
    and ensures that delete operations respect tenant_id constraints.
    """
    logger.debug("Running before_flush event listener")
    tenant_id = session.info.get("tenant_id")

    if tenant_id is not None:
        # Handle add operations
        for instance in session.new:
            if hasattr(instance, "tenant_id"):
                if getattr(instance, "tenant_id", None) is None:
                    setattr(instance, "tenant_id", tenant_id)
                    logger.debug(f"Set tenant_id for new instance: {instance}")
                elif getattr(instance, "tenant_id") != tenant_id:
                    # Most likely unnecessary since the select() prior to this will fail, due to adding the tenant_id of the user.
                    raise Exception(
                        f"Attempting to add instance with tenant_id that does not match session tenant_id {tenant_id}"
                    )

        # Handle delete operations
        for instance in session.deleted:
            if hasattr(instance, "tenant_id"):
                instance_tenant_id = getattr(instance, "tenant_id", None)
                if instance_tenant_id != tenant_id:
                    # Most likely unnecessary since the select() prior to this will fail, a 404 will be raised prior to this.
                    raise Exception(
                        f"Attempting to delete instance with tenant_id {instance_tenant_id} which does not match session tenant_id {tenant_id}"
                    )
                # Construct and execute a custom delete statement with tenant_id filter
                stmt = delete(instance.__class__).where(
                    instance.__class__.id == instance.id,
                )
                session.execute(stmt)
                session.expunge(
                    instance
                )  # Remove the instance from the session to prevent default delete
                logger.debug(
                    f"Modified delete statement to include tenant_id: {tenant_id}"
                )
```

Still new to SQLalchemy so forgive me if I misunderstood something. Appreciate the swift response.
